docs/architecture/adr-008-ics20-implementation.md

```diff
@@ -0,0 +1,435 @@
+        timeout_offset_timestamp: msg.timeout_timestamp
+    }
+}
+
+/// Handles incoming packets with ICS20 data
+/// To be called inside the on_recv_packet callback
+pub fn on_recv_packet<Ctx>(ctx: &Ctx, packet: Packet, data: FungibleTokenPacketData) -> Result<(), ICS20Error>
+    where Ctx: ICS20Context
+{
+    if !ctx.is_received_enabled() {
+        return Err(ICS20Error::receive_disabled());
+    }
+
+    let full_denom_path = data.denomination.to_string();
+    if receiver_chain_is_source(&packet.dest_port, &packet.dest_channel, full_denom_path.clone()) {
+        
+        // remove prefix added by sender chain
+        let voucher_prefix = get_denom_prefix(&packet.source_port, &packet.source_channel);
+        let unprefixed_denom = full_denom_path.clone().split_off(voucher_prefix.len());
+        
+        let denom = parse_denom_trace(&unprefixed_denom)?;
+        
+        let token = Coin {
+            denom,
+            amount: data.amount
+        };
+
+        let escrow_address = ctx.get_channel_escrow_address(packet.dest_port.clone(), packet.dest_channel.clone())?;
+        // Send tokens from escrow to receiver
+        ctx.send_coins(escrow_address, data.receiver.into())?;
+        
+        return Ok(())
+    }
+
+    let denom = get_denom_prefix(&packet.dest_port, &packet.dest_channel);
+    // NOTE: source_prefix contains the trailing "/"
+    denom.push_str(full_denom_path);
+    
+    let token = Coin {
+        denom: DenomTrace::parse_denom_trace(&denom)?,
+        amount: data.amount
+    };
+    
+    let module_acc = ctx.get_module_account();
+    // Mint vouchers to module
+    ctx.mint_coins(token.clone())?;
+    // Send vouchers to receiver
+    ctx.send_coins(module_acc, data.receiver.into())?;
+    Ok(())
+    
+}
+
+/// on_timeout_packet refunds the sender since the original packet sent was
+/// never received and has been timed out.
+/// To be called inside the on_timeout_packet callback
+pub fn on_timeout_packet<Ctx>(ctx: &Ctx, packet: Packet, data: FungibleTokenPacketData) -> Result<(), ICS20Error>
+    where Ctx: ICS20Context
+{
+    refund_packet_token(ctx, data)
+}
+
+/// on_acknowledgement_packet responds to the the success or failure of a packet
+/// acknowledgement written on the receiving chain. If the acknowledgement
+/// was a success then nothing occurs. If the acknowledgement failed, then
+/// the sender is refunded their tokens.
+/// To be called inside the on_acknowledgement_packet callback
+pub fn on_acknowledgement_packet<Ctx>(ctx: &Ctx, ack: ICS20Acknowledgement, data: FungibleTokenPacketData) -> Result<(), ICS20Error>
+    where Ctx: ICS20Context
+{
+    match ack {
+        ICS20Acknowledgement::Sucess => Ok(()),
+        _ => refund_packet_token(ctx, data)
+    }
+}
+
+/// Implements logic for refunding a sender on packet timeout or acknowledgement error
+pub fn refund_packet_token<Ctx>(ctx: &Ctx, data: FungibleTokenPacketData) -> Result<(), ICS20Error>
```

Comment:
Should be `data: &FungibleTokenPacketData`. This also applies to most other arguments in the ADR that are passed by value (e.g. all the `PortId` and `ChannelId` in `ICS20Keeper`).